app/carbonapi/http_handlers.go

```diff
@@ -428,37 +359,143 @@ func (app *App) renderHandler(w http.ResponseWriter, r *http.Request) {
+	from         string
+	until        string
+	format       string
+	template     string
+	useCache     bool
+	from32       int32
+	until32      int32
+	jsonp        string
+	cacheKey     string
+	cacheTimeout int32
+	qtz          string
+}
+
+func (app *App) renderHandlerProcessForm(r *http.Request, accessLogDetails *carbonapipb.AccessLogDetails, logger *zap.Logger) (renderForm, error) {
+	var res renderForm
+
+	err := r.ParseForm()
+	if err != nil {
+		return res, err
+	}
+
+	res.targets = r.Form["target"]
+	res.from = r.FormValue("from")
+	res.until = r.FormValue("until")
+	res.format = r.FormValue("format")
+	res.template = r.FormValue("template")
+	res.useCache = !parser.TruthyBool(r.FormValue("noCache"))
+
+	if res.format == jsonFormat {
+		// TODO(dgryski): check jsonp only has valid characters
+		res.jsonp = r.FormValue("jsonp")
+	}
+
+	if res.format == "" && (parser.TruthyBool(r.FormValue("rawData")) || parser.TruthyBool(r.FormValue("rawdata"))) {
+		res.format = rawFormat
+	}
+
+	if res.format == "" {
+		res.format = pngFormat
+	}
+
+	res.cacheTimeout = app.config.Cache.DefaultTimeoutSec
+
+	if tstr := r.FormValue("cacheTimeout"); tstr != "" {
+		t, err := strconv.Atoi(tstr)
+		if err != nil {
+			logger.Error("failed to parse cacheTimeout",
+				zap.String("cache_string", tstr),
+				zap.Error(err),
+			)
+		} else {
+			res.cacheTimeout = int32(t)
+		}
+	}
+
+	// make sure the cache key doesn't say noCache, because it will never hit
+	r.Form.Del("noCache")
+
+	// jsonp callback names are frequently autogenerated and hurt our cache
+	r.Form.Del("jsonp")
+
+	// Strip some cache-busters.  If you don't want to cache, use noCache=1
+	r.Form.Del("_salt")
+	r.Form.Del("_ts")
+	r.Form.Del("_t") // Used by jquery.graphite.js
+
+	res.cacheKey = r.Form.Encode()
+
+	// normalize from and until values
+	res.qtz = r.FormValue("tz")
+	res.from32 = date.DateParamToEpoch(res.from, res.qtz, timeNow().Add(-24*time.Hour).Unix(), app.defaultTimeZone)
+	res.until32 = date.DateParamToEpoch(res.until, res.qtz, timeNow().Unix(), app.defaultTimeZone)
+
+	accessLogDetails.UseCache = res.useCache
+	accessLogDetails.FromRaw = res.from
+	accessLogDetails.From = res.from32
+	accessLogDetails.UntilRaw = res.until
+	accessLogDetails.Until = res.until32
+	accessLogDetails.Tz = res.qtz
+	accessLogDetails.CacheTimeout = res.cacheTimeout
+	accessLogDetails.Format = res.format
+	accessLogDetails.Targets = res.targets
+
+	return res, nil
+}
+
+func (app *App) renderWriteBody(results []*types.MetricData, form renderForm, r *http.Request, logger *zap.Logger) ([]byte, error) {
 	var body []byte
 
-	switch format {
+	switch form.format {
 	case jsonFormat:
 		if maxDataPoints, _ := strconv.Atoi(r.FormValue("maxDataPoints")); maxDataPoints != 0 {
 			types.ConsolidateJSON(maxDataPoints, results)
 		}
 
 		body = types.MarshalJSON(results)
 	case protobufFormat, protobuf3Format:
-		body, err = types.MarshalProtobuf(results)
+		body, err := types.MarshalProtobuf(results)
```

Comment:
Won't this shadow the variable created in outer scope and create issues?

Reply:
This will shadow, but I don't see problem with that.

Reply:
https://play.golang.org/p/qkUigxU9eFN

Reply:
Yes, but what's the problem? Shadowing value is only used inside single case as expected.

Reply:
body would return nil in case format is one of protobufFormat, protobuf3Format

Reply:
If you talk about body, then yes, it is a problem.